python/tk_multi_workfiles/actions/file_action_factory.py

```diff
@@ -36,12 +37,39 @@
 from ..file_item import FileItem
 from ..user_cache import g_user_cache
 
+
 class FileActionFactory(object):
     """
+    Builder for the file open menu. Based on the current selection, the build will enumerate a list of actions,
+    grouped actions and seperators that should be displayed in the UI.
     """
-    
+
     def get_actions(self, file, work_area, file_model, workfiles_visible=True, publishes_visible=True):
         """
+        Retrives the list of actions for the given file.
+
+        :param file: FileItem instance representing the selection in the browser.
+        :param work_area: WorkArea instance representing the context associate with the file.
+        :param file_model: A FileModel instance.
+        :param workfiles_visible: True if workfiles are visible.
+        :param publishes_visible: True if publishes are visible.
+
+        :returns: List of Actions.
+        """
+        return self.__get_actions(file, work_area, file_model, workfiles_visible, publishes_visible, first_level=True)
+
+    def __get_actions(self, file, work_area, file_model, workfiles_visible, publishes_visible, first_level):
+        """
+        Retrives the list of actions for the given file, recursively.
```

Comment:
Maybe there's a better way to organize this method? It says it's recursive, but it's seemingly a fixed, single recursion to handle the previous version submenus. It's already a long method, so maybe that's another reason to break it up a bit and just have individual methods for returning the different types of actions rather than one big one with recursion. Just an idea. 

*extremely* simplified example:

```
def __get_actions(...):
  
    # ...
    actions.extend(self.__local_file_actions(...))
    actions.extend(self.__published_file_actions(...))
    actions.extend(self.__first_level_actions(...))
    actions.extend(self.__custom_actions(...))

    return actions

def __first_level_actions(...):

    # ...
    fl_actions.extend(self.__local_file_actions(...))
    fl_actions.extend(self.__published_file_actions(...))
    fl_actions.extedn(self.__custom_actions(...))
    
    return [group_action(fl_actions)]
```

Hope that makes sense.